### PR TITLE
Add controllers

### PR DESCRIPTION
**Pull request by arjunrn**
This PR adds the following in 3 commits:
- Initialize the operator with operator-sdk v0.16.0 and use operator-custom-metrics for a `ServiceMonitor`
- Bootstrap boilerplate
- Add controllers with tests.

### CODE REVIEW

File: pkg/metrics/aggregator.go
Comment (2uasimojo) [inline pkg/metrics/aggregator.go:32]: nit: s/a/an/

File: pkg/metrics/aggregator.go
Comment (2uasimojo) [inline pkg/metrics/aggregator.go:90]: nit: you could unlock here

File: cmd/manager/main.go
Comment (2uasimojo) [inline cmd/manager/main.go:129]: nit: for readability, this could be closer to the metrics config setup (L154)

File: deploy/role.yaml
Comment (2uasimojo) [inline deploy/role.yaml:1]: This could probably be pared down dramatically.

**Review by 2uasimojo — APPROVED** [review]
Overall this seems like a fine start. Other than which bits go in which commits (which isn't a show-stopper) everything could be addressed in a followon. I'm going to

/lgtm

this, but

/hold

it in case you want to fix stuff up in this PR. (Not that any of that matters right now, since you don't have prow set up yet.)

File: deploy/cluster_role.yaml
Comment (arjunrn) [reply deploy/cluster_role.yaml:25]: `create` can be dropped. But the others are required. `get` and `list` are needed by the `controller-runtime` itself.

File: pkg/controller/add_controllers.go
Comment (arjunrn) [reply pkg/controller/add_controllers.go:1]: Whenever a new controller is added a new file called `add_$controller_name.go` is added to `pkg/controllers`. Since there is new file for every controller which is kind of too much I decided to aggregate them into a single file called `add_controllers.go`. So this is not part of the initial generated operator skeleton.

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (arjunrn) [reply pkg/controller/clusterrole/clusterrole_controller.go:1]: So you meant I could have generated the controllers as well along with the operator stuff. And then filled it out in this commit. Well, I could still do it but it will take some work. Can you review this as is for now.

File: pkg/metrics/aggregator.go
Comment (arjunrn) [reply pkg/metrics/aggregator.go:90]: Yes, but the `defer mutex.Unlock()` statement after a `Lock()` is idiomatic go. The reason for it is that the deferred statement will always be executed even if there is an error or panic somewhere in the call stack. So you can be sure there are no deadlocks. I could move  map update into a separate function if you think the lock is being held for too long but I'm quite sure it's not too much overhead.

File: cmd/manager/main.go
Comment (arjunrn) [reply cmd/manager/main.go]: Correct

File: deploy/cluster_role.yaml
Comment (arjunrn) [reply deploy/cluster_role.yaml]: Removed `create`

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (arjunrn) [reply pkg/controller/clusterrole/clusterrole_controller.go]: Added predicates.

File: pkg/metrics/metrics.go
Comment (arjunrn) [reply pkg/metrics/metrics.go]: I've changed it to a singleton class.

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (2uasimojo) [reply pkg/controller/clusterrole/clusterrole_controller.go:1]: Yeah totally, don't worry about it, just for future.

File: pkg/controller/add_controllers.go
Comment (2uasimojo) [reply pkg/controller/add_controllers.go:1]: Ack.

File: deploy/cluster_role.yaml
Comment (2uasimojo) [reply deploy/cluster_role.yaml]: Cool.

Yeah, the read verbs are needed for `Watch`ing, I get that.

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (2uasimojo) [inline pkg/controller/clusterrole/clusterrole_controller.go:26]: "adoption" seems like an odd choice. Where did that come from?

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (2uasimojo) [inline pkg/controller/clusterrole/clusterrole_controller.go:67]: Noyce.

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (2uasimojo) [inline pkg/controller/clusterrole/clusterrole_controller.go:117]: You don't need `containsString` -- `AddFinalizer` already does that check for you. Similar throughout.

File: pkg/metrics/aggregator.go
Comment (2uasimojo) [inline pkg/metrics/aggregator.go:29]: It would be nice if `NewMetricsAggregator` could be private. But I get that you need to use it for test to reduce the aggregation interval. Could you maybe put that in a docstring comment on the func? Including "Don't use this outside of test!"

File: pkg/metrics/aggregator.go
Comment (2uasimojo) [reply pkg/metrics/aggregator.go:90]: Yeah, don't worry about it; prefer using the idiom, and not worth splitting up the function.

**Review by 2uasimojo — COMMENTED** [review]
Getting close. Again, nothing blocking here.

File: pkg/controller/clusterrole/clusterrole_controller.go
Comment (arjunrn) [reply pkg/controller/clusterrole/clusterrole_controller.go:117]: Without the condition the controller would update the finalizer in every reconcile loop.

**Comment by 2uasimojo** [comment]
/lgtm
/approve

If you want to fix that comment typo, you don't have to wait for another approval.

May want to squash the last three commits, though.

**Pull request merged**
